core-kafka-streams/src/main/java/org/galatea/kafka/starter/messaging/streams/TaskStore.java

```diff
@@ -0,0 +1,51 @@
+package org.galatea.kafka.starter.messaging.streams;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.galatea.kafka.starter.messaging.streams.util.RetentionPolicy;
+
+@RequiredArgsConstructor(access = AccessLevel.PACKAGE)
+public class TaskStore<K, V> implements KafkaStreamsStore<K, V> {
+
+  private final KeyValueStore<K, V> inner;
+  private final RetentionPolicy<K, V> retentionPolicy;
+  private final TaskContext taskContext;
+
+  @Override
+  public Optional<V> get(K key) {
+    V raw = inner.get(key);
+    if (raw != null && retentionPolicy != null
+        && !retentionPolicy.shouldKeep(key, raw, taskContext)) {
```

Comment:
Is this sort of like "self-cleaning"?

Reply:
I would classify it as "on-the-fly" cleaning :)